### ISSUE TITLE
Git flow manifest management

### ISSUE THREAD

**Issue by smipi1** [GatorQue/git-reflow#9]
Along with the git flow project branch management support, we need to add the manifest management functionality so that repo flow truly works at the composition (manifest) level. Before coming up with a design, we need to clarify the requriements:
1. When a feature / release / hotfix / support is started, a tracking manifest must be locally made available, so that the developer can later resume the work.
2. When a feature / release / hotfix is published, the tracking manifest should also be published to the remote manifest repository.
3. When a feature / release / hotfix is tracked, the tracking manifest for the feature should be checked out.
4. When a feature / release / hotfix is finished, the tracking manifest should cease to exist.
5. When a feature is finished, the develop manifest should be updated with any changes (this may include manual changes to the manifest structure, revisions, etc).
6. When a release / hotfix is finished, the develop and master manifests should be updated with any changes (this may include manual changes to the manifest structure, revisions, etc).
7. When the develop or master manifest is updated on a feature / release / hotfix finish, upstream changes may not be clobbered (We should be generating merge conflicts).
8. A feature / release / hotfix / support may not be started if there are local uncommitted changes.

**Comment by smipi1** [comment]
@GatorQue,

I have been banging my head against the best way to manage feature, hotfix, release and support manifests with `repo flow`. I would like to have all WIP (maifests) in the manifest repository on the master branch. This would allow you to use `repo diffmanifests` and similar tools, but has some less than ideal consequences:
1. A `repo flow feature start` effectively is a fork of `develop.xml`: `develop.xml` gets copied to `feature/<name>.xml`, `git flow` creates a `feature/<name>` branch in each of the flow-enabled projects and the `revision` attribute in `feature/<name>.xml` is modified to track the `git flow` branch.
2. A `repo flow feature finish` effectively merges the fork back into `develop.xml`: `git flow` merges the branch back into `develop` in each flow-enabled project, the `revision` attribute in `feature/<name>.xml` is modified to track the `develop` branch again, and then `feature/<name>.xml` is moved to replace `develop.xml`. But what should happen if the upstream `develop.xml` was also changed by someone else?
   - We should generate merge conflicts where appropriate.
   - This will not happen by itself, and you cannot expect git to generate the appropriate behavior if everything is done on the same `master` branch in the manifests repository.
3. A `repo flow feature publish` should expose the feature manifest and associated project branches to other developers via remote, but it should expose ONLY that. With a single manifest repository branch, this will not be the case when a developer is working on multiple features, hotfixes and releases concurrently and decides to publish only one.
4. A `repo flow feature delete` should discard the feature manifest and associated project branches as if it never happened. This will not be the case when everything happens on a manifest `master` branch.

It is starting to look like a branch per feature, release or hotfix in the manifest repository is the easiest way to get most of the behavior correct. The only pain point, is that the `repo diffmanifests` command only supports comparing manifests currently checked out (I.e. visible on the same branch within the manifests repository).

Have you got any suggestions?

**Comment by GatorQue** [comment]
I agree that branching might be the only way to go with handling the manifest repository. But I'm not sure why every feature.xml file would need to be pushed when a `repo flow feature publish` command was executed. If `repo flow feature start` does a "repo switch" to the feature/feature-name.xml, it should already know which file to publish and can _only_ publish this file and any other files it may include/reference.  But this causes issues if those same files are being referenced (but not changed) by other features. Therefore we are back to a branching model for the manifest repository.
My recommendation is to improve the repo diffmanifests command to support comparisons across branches, or create a repo flow diff command to show manifest differences for the current branch against the "develop" branch. Do you have a preference on which way we go (modify repo diffmanifests or create some type of repo flow diff command)?

**Comment by smipi1** [comment]
@GatorQue,
Carefully reading your comment gives me the impression that you may already understand what I was referring to: Should we decide to keep everything flat in the manifest repository (no branches), publishing a feature would result in the entire committed state of the manifest repository being pushed. That would include all WIP (feature, hotfix, release and support manifests). This certainly is less than ideal.

As you already hinted, we could decide not to commit WIP in the manifest repository, unless it has to be published. I expect that in addition to the included manifest problem, we will be playing whack-a-mole with any number of issues arising from this.

So let's assume that a branching model in the manifest repository is the best approach. Why should we then go for a hybrid approach? If we stick to always having a `default.xml`, and stick to using only branches to fork and merge the composition, we can apply git-flow to the manifest repository as well. We will not have to come up with anything bespoke.

We can improve `repo diffmanifests` so that we can specify branches / tags, but it was never intended for this use-case.

If we go for a full-blown git-flow branching model in the manifest repository, we will be better off adding this to `repo flow`. We already have a `repo flow feature diff` that should compare with `develop`. We can add a `repo flow diff` with the ability to compare any flow-related ref:

```
repo flow diff -d -m # compare develop and master
repo flow diff -d -f <name> # compare develop and feature called <name>
# etc.
```

Or something like this:

```
repo flow feature diff -d <name> # compare feature called <name> with develop
repo flow feature diff <name1> <name2> # compare two features
repo flow feature diff <name> -r <version> # compare feature with release
repo flow diff -r <version1> -r <version2>
```

**Comment by GatorQue** [comment]
I agree with that approach. Another use case that this would help with is the ability to "throw away" a feature/release/hotfix/ branch without affecting the other files. This would become easier if using a branching approach. I like the idea of keeping the user editing the manifest repository manually to a minimum (as needed).
